### ISSUE TITLE
Llvm calling libsyntax causing panic

### ISSUE THREAD

**Issue by bjorn3** [rust-lang/rust#36745]
At irc pounce_ got a strange compiler panic:
https://gist.github.com/4e554c4c/0a99cb7c0f87e352538c1a4755e6316f

**Comment by sanxiyn** [comment]
How can one reproduce this?

**Comment by bjorn3** [comment]
pounce_ tried to compile the code of http://os.phil-opp.com/better-exception-messages.html .

**Comment by 4e554c4c** [comment]
The error is caused after adding an inline assembly macro to the code that passes a struct that is on the stack to a function. 
```Rust
macro_rules! handler {
    ($name: ident) => {{
        #[naked]
        extern "C" fn wrapper() -> ! {
            unsafe {
                asm!("mov rdi, rsp
                      sub rsp, 8 // Align stack pointer
                      call $0"
                      :: "i"($name as extern "C" fn(_) -> !)
                      : "rdi" : "intel");
                ::core::intrinsics::unreachable();
            }
        }
        wrapper
    }}
}
```
I use the macro with a `pub extern "C" fn` that takes a struct pointer as input. Currently unable to reproduce, suggestions are welcome.

**Comment by bjorn3** [comment]
target spec: https://gist.github.com/4e554c4c/1c3839545631308b49faaa2e5b088774

**Comment by 4e554c4c** [comment]
I have pushed the code that causes this error to my git repository. The file that causes the problem can be found [here](https://github.com/ESALP/ESALP-1/blob/rustc_error/src/interrupts/mod.rs). That branch can compile, one needs a nightly rust compiler and xargo installed to do so however. Sorry about all the things that are commented out, I have tried to limit what compiles to limit noise. The file used to link to an assembly file, but now it only links to two functions: `sti()` , which is simply:
```asm
sti:
    sti
    ret
```
and `KEXIT()` , which is my kernel exit point. The problem is fairly contained in that one file.
